mirrord-agent/src/tcp/outgoing.rs

```diff
@@ -0,0 +1,177 @@
+                                    let connect_response =
+                                        TcpStream::connect(remote_address)
+                                            .await
+                                            .map_err(From::from)
+                                            .map(|remote_stream| {
+                                                let connection_id = writers
+                                                    .keys()
+                                                    .last()
+                                                    .copied()
+                                                    .map(|last| last + 1)
+                                                    .unwrap_or_default();
+
+                                                // Split the `remote_stream` so we can keep reading
+                                                // and writing from multiple hosts without blocking.
+                                                let (read_half, write_half) = remote_stream.into_split();
+                                                writers.insert(connection_id, write_half);
+                                                readers.insert(connection_id, ReaderStream::new(read_half));
+
+                                                ConnectResponse {
+                                                    connection_id,
+                                                    remote_address,
+                                                }
+                                            });
+
+                                    let response = TcpOutgoingResponse::Connect(connect_response);
+                                    response_tx.send(response).await?
+                                }
+                                // [user] -> [layer] -> [agent] -> [remote]
+                                // `user` wrote some message to the remote host.
+                                TcpOutgoingRequest::Write(WriteRequest {
+                                    connection_id,
+                                    bytes,
+                                }) => {
+                                    let write_response = match writers
+                                        .get_mut(&connection_id)
+                                        .ok_or(ResponseError::NotFound(connection_id as usize))
+                                    {
+                                        Ok(writer) => writer
+                                            .write_all(&bytes)
+                                            .await
+                                            .map_err(ResponseError::from)
+                                            .map(|()| WriteResponse { connection_id }),
+                                        Err(fail) => Err(fail),
+                                    };
+
+                                    let response = TcpOutgoingResponse::Write(write_response);
```

Comment:
why do you write back write_response? if it fails to write socket will be terminated anyway?

Reply:
We can get away without it, as you pointed out to me, but as I had already implemented this message flow, why remove it?